browser/static/js/adjust.js

```diff
@@ -1,277 +1,398 @@
+
+    // brightness and contrast adjustment
+    this._minContrast = -100;
+    this._maxContrast = 700;
+
+    this._minBrightness = -512;
+    this._maxBrightness = 255;
+
+    this.contrastMap = new Map();
+    this.brightnessMap = new Map();
+
+    for (let i = 0; i < channelMax; i++) {
+      this.brightnessMap.set(i, 0);
+      this.contrastMap.set(i, 0);
+    }
+    this.brightness = this.brightnessMap.get(0);
+    this.contrast = this.contrastMap.get(0);
+
+    // raw and adjusted image storage
+    // cascasding image updates if raw or seg is reloaded
+    this.rawImage = new Image();
+    this.contrastedRaw = new Image();
+    this.preCompRaw = new Image();
+
+    this.segImage = new Image();
+    this.preCompSeg = new Image();
+
+    // adjusted raw + annotations
+    this.compositedImg = new Image();
+
+    // composite image + outlines, transparent highlight
+    this.postCompImg = new Image();
+
+    // TODO: undefined variables
+    if (rgb) {
+      this.rawImage.onload = () => this.contrastRaw();
+      this.contrastedRaw.onload = () => this.rawAdjust(seg_array, current_highlight, edit_mode, brush, mode);
+      this.segImage.onload = () => this.preCompAdjust(seg_array, current_highlight, edit_mode, brush, mode);
+      this.preCompSeg.onload = () => this.segAdjust(seg_array, current_highlight, edit_mode, brush, mode);
+    } else {
+      this.rawImage.onload = () => this.contrastRaw();
+      this.contrastedRaw.onload = () => this.preCompRawAdjust();
+      this.preCompRaw.onload = () => this.rawAdjust(seg_array, current_highlight, edit_mode, brush, mode);
+      this.segImage.onload = () => this.preCompAdjust(seg_array, current_highlight, edit_mode, brush, mode);
+      this.preCompSeg.onload = () => this.segAdjust(seg_array, current_highlight, edit_mode, brush, mode);
+      this.compositedImg.onload = () => this.postCompAdjust(seg_array, edit_mode, brush);
+    }
 
-// image adjustment functions: take img as input and manipulate data attribute
-// pixel data is 1D array of 8bit RGBA values
-function contrast_image(img, contrast = 0, brightness = 0) {
-  let d = img.data;
-  contrast = (contrast / 100) + 1;
-  for (let i = 0; i < d.length; i += 4) {
-      d[i] = d[i]*contrast + brightness;
-      d[i + 1] = d[i+1]*contrast + brightness;
-      d[i + 2] = d[i+2]*contrast + brightness;
+    this.rawLoaded = false;
+    this.segLoaded = false;
+  }
+
+  // getters for brightness/contrast allowed ranges
+  // no setters; these should remain fixed
+  get minBrightness() {
+    return this._minBrightness;
+  }
+
+  get maxBrightness() {
+    return this._maxBrightness;
+  }
+
+  get minContrast() {
+    return this._minContrast;
+  }
+
+  get maxContrast() {
+    return this._maxContrast;
+  }
+
+  // TODO: getters/setters for brightness and contrast?
+  // or use this group of methods?
+
+  changeContrast(inputChange) {
+    const modContrast = -Math.sign(inputChange) * 4;
+    // stop if fully desaturated
```

Comment:
I think this TODO has been resolved by the getters above.

```suggestion
  changeContrast(inputChange) {
    const modContrast = -Math.sign(inputChange) * 4;
    // stop if fully desaturated
```